NullPointerException

```
java.lang.NullPointerException: Cannot invoke "org.aya.syntax.literate.CodeOptions$NormalizeMode.ordinal()" because "<parameter2>" is null
        at aya.base/org.aya.normalize.Normalizer.normalize(Unknown Source)
        at aya.cli.impl/org.aya.cli.interactive.ReplCompiler.lambda$compileToContext$7(Unknown Source)
        at kala.base@0.73.0/kala.control.Either.map(Unknown Source)
        at aya.cli.impl/org.aya.cli.interactive.ReplCompiler.compileToContext(Unknown Source)
        at aya.cli.impl/org.aya.cli.interactive.ReplCompiler.compileToContext(Unknown Source)
        at aya.cli.console/org.aya.cli.repl.AyaRepl.eval(Unknown Source)
        at aya.repl/org.aya.repl.Repl.loop(Unknown Source)
        at aya.cli.console/org.aya.cli.repl.AyaRepl.singleLoop(Unknown Source)
```

```
Caused by: java.lang.IndexOutOfBoundsException: Index: 1 Size: 1
	at kala.collection@0.73.0/kala.collection.immutable.ImmutableVectors$Vector1.get(Unknown Source)
	at aya.syntax/org.aya.syntax.compile.JitTele$LocallyNameless.telescope(Unknown Source)
	at aya.syntax/org.aya.syntax.compile.JitTele.telescope(Unknown Source)
	at aya.base/org.aya.unify.TermComparator.compareMany(Unknown Source)
	at aya.base/org.aya.unify.TermComparator.compareCallApprox(Unknown Source)
	at aya.base/org.aya.unify.TermComparator.compareApprox(Unknown Source)
	at aya.base/org.aya.unify.TermComparator.compare(Unknown Source)
	at aya.base/org.aya.tyck.tycker.Unifiable.unifyTerm(Unknown Source)
	at aya.base/org.aya.tyck.tycker.Unifiable.unifyTermReported(Unknown Source)
```

--- a/base/src/main/java/org/aya/normalize/Normalizer.java
+++ b/base/src/main/java/org/aya/normalize/Normalizer.java
@@ -16,7 +16,7 @@ import org.aya.syntax.core.term.marker.BetaRedex;
 import org.aya.syntax.core.term.marker.StableWHNF;
 import org.aya.syntax.core.term.xtt.CoeTerm;
 import org.aya.syntax.core.term.xtt.DimTerm;
-import org.aya.syntax.literate.CodeOptions;
+import org.aya.syntax.literate.CodeOptions.NormalizeMode;
 import org.aya.syntax.ref.AnyVar;
 import org.aya.tyck.TyckState;
 import org.aya.tyck.tycker.Stateful;
@@ -132,11 +132,12 @@ public final class Normalizer implements UnaryOperator<Term> {
    * Do NOT use this in the type checker.
    * This is for REPL/literate mode and testing.
    */
-  public @NotNull Term normalize(Term term, CodeOptions.NormalizeMode mode) {
+  public @NotNull Term normalize(Term term, NormalizeMode mode) {
     return switch (mode) {
       case HEAD -> apply(term);
       case FULL -> new Full().apply(term);
       case NULL -> new Finalizer.Freeze(() -> state).zonk(term);
+      case null -> new Finalizer.Freeze(() -> state).zonk(term);
     };
   }
 }

--- a/base/src/main/java/org/aya/normalize/PatMatcher.java
+++ b/base/src/main/java/org/aya/normalize/PatMatcher.java
@@ -13,7 +13,6 @@ import org.aya.syntax.core.term.TupTerm;
 import org.aya.syntax.core.term.call.ConCall;
 import org.aya.syntax.core.term.call.ConCallLike;
 import org.aya.syntax.core.term.repr.IntegerTerm;
-import org.aya.syntax.core.term.repr.ListTerm;
 import org.aya.tyck.pat.BindEater;
 import org.aya.util.error.Panic;
 import org.jetbrains.annotations.NotNull;
@@ -81,7 +80,7 @@ public record PatMatcher(boolean inferMeta, @NotNull UnaryOperator<Term> pre) {
 
   /**
    * @return a substitution of corresponding bindings of {@param pats} if success.
-   * @apiNote The binding order is the same as {@link Pat#consumeBindings(java.util.function.BiConsumer)}
+   * @apiNote The binding order is the same as {@link Pat#collectVariables}
    * @see State
    */
   public @NotNull Result<ImmutableSeq<Term>, State> apply(

--- a/base/src/main/java/org/aya/terck/CallResolver.java
+++ b/base/src/main/java/org/aya/terck/CallResolver.java
@@ -144,8 +144,8 @@ public record CallResolver(
 
   @Override public void accept(@NotNull Term.Matching matching) {
     this.currentClause.set(matching);
-    var vars = Pat.collectBindings(matching.patterns().view()).view().map(Pat.CollectBind::var);
-    visitTerm(matching.body().instantiateTeleVar(vars));
+    var vars = Pat.collectVariables(matching.patterns().view()).component1();
+    visitTerm(matching.body().instantiateTeleVar(vars.view()));
     this.currentClause.set(null);
   }
 

--- a/base/src/main/java/org/aya/tyck/StmtTycker.java
+++ b/base/src/main/java/org/aya/tyck/StmtTycker.java
@@ -172,13 +172,11 @@ public record StmtTycker(
       lhsResult.addLocalLet(ownerBinds, tycker);
       freeDataCall = new DataCall(dataRef, 0, wellPats.map(PatToTerm::visit));
 
-      var patsWithTermBound = Pat.collectVariables(wellPats.view());
-      wellPats = patsWithTermBound.component2();
-      var allBinds = Pat.collectBindings(wellPats.view());
-      ownerBinds = patsWithTermBound.component1().toImmutableSeq();
-      TeleTycker.bindTele(ownerBinds, allBinds);
+      var allTypedBinds = Pat.collectBindings(wellPats.view());
+      ownerBinds = lhsResult.allBinds();
+      TeleTycker.bindTele(ownerBinds, allTypedBinds);
       ownerTele = ownerBinds
-        .zip(allBinds, (bind, param) -> new WithPos<>(bind.definition(), param));
+        .zip(allTypedBinds, (bind, param) -> new WithPos<>(bind.definition(), param));
       if (wellPats.allMatch(pat -> pat instanceof Pat.Bind))
         wellPats = ImmutableSeq.empty();
     } else {

--- a/base/src/main/java/org/aya/tyck/StmtTycker.java
+++ b/base/src/main/java/org/aya/tyck/StmtTycker.java
@@ -3,7 +3,6 @@
 package org.aya.tyck;
 
 import kala.collection.immutable.ImmutableSeq;
-import kala.collection.mutable.MutableSeq;
 import kala.control.Either;
 import kala.control.Option;
 import org.aya.generic.Modifier;
@@ -173,14 +172,13 @@ public record StmtTycker(
       lhsResult.addLocalLet(ownerBinds, tycker);
       freeDataCall = new DataCall(dataRef, 0, wellPats.map(PatToTerm::visit));
 
+      var patsWithTermBound = Pat.collectVariables(wellPats.view());
+      wellPats = patsWithTermBound.component2();
       var allBinds = Pat.collectBindings(wellPats.view());
-      ownerBinds = allBinds.map(Pat.CollectBind::var);
-      var preOwnerTele = MutableSeq.from(allBinds.view()
-        .map(x -> new Param(x.var().name(), x.type(), false)));
-      TeleTycker.bindTele(ownerBinds, preOwnerTele);
-      ownerTele = allBinds
-        .zip(preOwnerTele, (bind, param) -> new WithPos<>(bind.var().definition(), param))
-        .toImmutableSeq();
+      ownerBinds = patsWithTermBound.component1().toImmutableSeq();
+      TeleTycker.bindTele(ownerBinds, allBinds);
+      ownerTele = ownerBinds
+        .zip(allBinds, (bind, param) -> new WithPos<>(bind.definition(), param));
       if (wellPats.allMatch(pat -> pat instanceof Pat.Bind))
         wellPats = ImmutableSeq.empty();
     } else {

--- a/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
+++ b/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
@@ -192,7 +192,7 @@ public record ClauseTycker(@NotNull ExprTycker exprTycker) implements Problemati
         wellBody = exprTycker.inherit(bodyExpr, result.type).wellTyped();
 
         // bind all pat bindings
-        var patBindTele = Pat.collectBindings(result.clause.pats().view()).view().map(Pat.CollectBind::var);
+        var patBindTele = Pat.collectVariables(result.clause.pats().view()).component1().view();
         wellBody = wellBody.bindTele(patBindTele);
       }
 

--- a/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
+++ b/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
@@ -45,6 +45,7 @@ public record ClauseTycker(@NotNull ExprTycker exprTycker) implements Problemati
   public record LhsResult(
     @NotNull LocalCtx localCtx,
     @NotNull Term type,
+    @NotNull ImmutableSeq<LocalVar> allBinds,
     @NotNull ImmutableSeq<Jdg> paramSubst,
     @NotNull LocalLet asSubst,
     @NotNull Pat.Preclause<Expr> clause,
@@ -157,10 +158,11 @@ public record ClauseTycker(@NotNull ExprTycker exprTycker) implements Problemati
       // * telescope are well-typed and no Meta
       // * PatternTycker doesn't introduce any Meta term
       ctx = ctx.map(TermInline::apply);
+      var patWithTypeBound = Pat.collectVariables(patResult.wellTyped().view());
 
-      var newClause = new Pat.Preclause<>(clause.sourcePos, patResult.wellTyped(), patResult.newBody());
-      return new LhsResult(ctx, resultTerm, patResult.paramSubst(),
-        patResult.asSubst(), newClause, patResult.hasError());
+      var newClause = new Pat.Preclause<>(clause.sourcePos, patWithTypeBound.component2(), patResult.newBody());
+      return new LhsResult(ctx, resultTerm, patWithTypeBound.component1().toImmutableSeq(),
+        patResult.paramSubst(), patResult.asSubst(), newClause, patResult.hasError());
     });
   }
 

--- a/base/src/main/java/org/aya/tyck/pat/PatUnify.java
+++ b/base/src/main/java/org/aya/tyck/pat/PatUnify.java
@@ -1,0 +1,111 @@
+// Copyright (c) 2020-2024 Tesla (Yinsen) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.tyck.pat;
+
+import kala.collection.Seq;
+import kala.collection.SeqView;
+import kala.collection.immutable.ImmutableSeq;
+import kala.collection.mutable.MutableList;
+import org.aya.syntax.core.pat.Pat;
+import org.aya.syntax.core.pat.PatToTerm;
+import org.aya.syntax.core.term.FreeTerm;
+import org.aya.syntax.core.term.Term;
+import org.aya.syntax.core.term.TupTerm;
+import org.aya.syntax.core.term.call.ConCall;
+import org.aya.syntax.ref.LocalCtx;
+import org.aya.syntax.ref.LocalVar;
+import org.aya.util.error.Panic;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * The unification of patterns. This is <strong>not</strong> pattern unification.
+ *
+ * @author ice1000
+ * @see #unifyPat
+ */
+public record PatUnify(
+  @NotNull MutableList<Term> lhsSubst, @NotNull MutableList<Term> rhsSubst,
+  @NotNull LocalCtx ctx
+) {
+  public record Result(PatUnify unify, @NotNull Seq<Term> args) { }
+
+  private @NotNull Term unify(@NotNull Pat lhs, @NotNull Pat rhs) {
+    return switch (lhs) {
+      case Pat.Bind bind -> visitAs(bind.bind(), rhs);
+      case Pat.Tuple(var elements) -> rhs instanceof Pat.Tuple(var erements)
+        ? new TupTerm(visitList(elements, erements))
+        : Panic.unreachable();
+      case Pat.Con con -> switch (rhs) {
+        case Pat.Con con1 -> {
+          // Assumption
+          assert con.ref() == con1.ref();
+          yield new ConCall(con.conHead(), visitList(con.args(), con1.args()));
+        }
+        case Pat.ShapedInt lit -> unify(con, lit.constructorForm());
+        default -> Panic.unreachable();
+      };
+      case Pat.ShapedInt lhsInt -> switch (rhs) {
+        case Pat.ShapedInt _ -> lhsInt.toTerm();
+        case Pat.Con con -> unify(lhsInt.constructorForm(), con);
+        // Try one more time in case we add more rhs case when lhs is a constructor.
+        // see: PatMatcher#match(Pat, Term)
+        default -> unify(lhsInt.constructorForm(), rhs);
+      };
+      default -> Panic.unreachable();
+    };
+  }
+
+  private @NotNull ImmutableSeq<Term> visitList(ImmutableSeq<Pat> lpats, ImmutableSeq<Pat> rpats) {
+    assert rpats.sizeEquals(lpats.size());
+    return lpats.zip(rpats, (lp, rp) -> unifyPat(lp, rp, ctx, lhsSubst, rhsSubst));
+  }
+
+  private @NotNull Term visitAs(@NotNull LocalVar as, Pat rhs) {
+    if (rhs instanceof Pat.Bind(var bind, var ty)) {
+      var fresh = new FreeTerm(new LocalVar(as.name() + bind.name()));
+      ctx.put(fresh.name(), ty.instantiateTele(rhsSubst.view()));
+      lhsSubst.append(fresh);
+      rhsSubst.append(fresh);
+      return fresh;
+    } else {
+      rhs.consumeBindings((v, ty) -> {
+        ctx.put(v, ty.instantiateTele(rhsSubst.view()));
+        rhsSubst.append(new FreeTerm(v));
+      });
+      var e = PatToTerm.visit(rhs).instantiateTele(rhsSubst.view());
+      lhsSubst.append(e);
+      return e;
+    }
+  }
+
+  private static @NotNull Term unifyPat(
+    Pat lhs, Pat rhs, LocalCtx ctx,
+    MutableList<Term> lhsSubst, MutableList<Term> rhsSubst
+  ) {
+    if (rhs instanceof Pat.Bind) {
+      var unify = new PatUnify(rhsSubst, lhsSubst, ctx);
+      return unify.unify(rhs, lhs);
+    } else {
+      var unify = new PatUnify(lhsSubst, rhsSubst, ctx);
+      return unify.unify(lhs, rhs);
+    }
+  }
+
+  /**
+   * The unification of patterns. Assumes well-typedness, homogeneous-ness and positive success.
+   *
+   * @param lhsSubst the substitutions that would turn the lhs pattern to the rhs one.
+   * @param rhsSubst the substitutions that would turn the rhs pattern to the lhs one.
+   * @return a ctx that contains all variables that are not unified.
+   * @throws Panic if failed
+   * @see PatUnify#visitAs(LocalVar, Pat)
+   */
+  public static @NotNull Result unifyPat(
+    @NotNull SeqView<Pat> lpats, @NotNull SeqView<Pat> rpats, @NotNull LocalCtx ctx,
+    @NotNull MutableList<Term> lhsSubst, @NotNull MutableList<Term> rhsSubst
+  ) {
+    assert rpats.sizeEquals(lpats);
+    var args = lpats.zip(rpats, (lp, rp) -> unifyPat(lp, rp, ctx, lhsSubst, rhsSubst));
+    return new Result(new PatUnify(lhsSubst, rhsSubst, ctx), args.toImmutableSeq());
+  }
+}

--- a/base/src/main/java/org/aya/tyck/pat/PatUnify.java
+++ b/base/src/main/java/org/aya/tyck/pat/PatUnify.java
@@ -38,7 +38,7 @@ public record PatUnify(
       case Pat.Con con -> switch (rhs) {
         case Pat.Con con1 -> {
           // Assumption
-          assert con.ref() == con1.ref();
+          assert con.ref().equals(con1.ref());
           yield new ConCall(con.conHead(), visitList(con.args(), con1.args()));
         }
         case Pat.ShapedInt lit -> unify(con, lit.constructorForm());

--- a/base/src/main/java/org/aya/tyck/pat/PatUnify.java
+++ b/base/src/main/java/org/aya/tyck/pat/PatUnify.java
@@ -68,12 +68,12 @@ public record PatUnify(
       rhsSubst.append(fresh);
       return fresh;
     } else {
+      var e = PatToTerm.visit(rhs).instantiateTele(rhsSubst.view());
+      lhsSubst.append(e);
       rhs.consumeBindings((v, ty) -> {
         ctx.put(v, ty.instantiateTele(rhsSubst.view()));
         rhsSubst.append(new FreeTerm(v));
       });
-      var e = PatToTerm.visit(rhs).instantiateTele(rhsSubst.view());
-      lhsSubst.append(e);
       return e;
     }
   }

--- a/base/src/main/java/org/aya/tyck/pat/YouTrack.java
+++ b/base/src/main/java/org/aya/tyck/pat/YouTrack.java
@@ -2,13 +2,15 @@
 // Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
 package org.aya.tyck.pat;
 
+import kala.collection.Seq;
 import kala.collection.immutable.ImmutableSeq;
 import kala.collection.mutable.MutableLinkedSet;
+import kala.collection.mutable.MutableList;
 import kala.collection.mutable.MutableSet;
 import kala.control.Option;
 import org.aya.normalize.PatMatcher;
 import org.aya.syntax.core.pat.Pat;
-import org.aya.syntax.core.pat.PatToTerm;
+import org.aya.syntax.core.term.FreeTerm;
 import org.aya.syntax.core.term.Param;
 import org.aya.syntax.core.term.Term;
 import org.aya.syntax.ref.LocalCtx;
@@ -37,12 +39,15 @@ public record YouTrack(
     MutableSet<ClausesProblem.Domination> doms
   ) {
     var ctx = tycker.localCtx().derive();
-    var args = new PatToTerm.Binary(ctx).list(
-      lhsInfo.matching.patterns(), rhsInfo.matching.patterns());
-    domination(ctx, args, lhsInfo.ix, rhsInfo.ix, rhsInfo.matching, doms);
-    domination(ctx, args, rhsInfo.ix, lhsInfo.ix, lhsInfo.matching, doms);
-    var lhsTerm = prebuiltMatcher.apply(lhsInfo.matching, args).get();
-    var rhsTerm = prebuiltMatcher.apply(rhsInfo.matching, args).get();
+    var unifyResult = PatUnify.unifyPat(
+      lhsInfo.matching.patterns().view(),
+      rhsInfo.matching.patterns().view(), ctx,
+      MutableList.create(), MutableList.create());
+    var unify = unifyResult.unify();
+    domination(ctx, unify.rhsSubst(), lhsInfo.ix, rhsInfo.ix, rhsInfo.matching, doms);
+    domination(ctx, unify.lhsSubst(), rhsInfo.ix, lhsInfo.ix, lhsInfo.matching, doms);
+    var lhsTerm = lhsInfo.matching.body().instantiateTele(unify.lhsSubst().view());
+    var rhsTerm = rhsInfo.matching.body().instantiateTele(unify.rhsSubst().view());
     // // TODO: Currently all holes at this point are in an ErrorTerm
     // if (lhsTerm instanceof ErrorTerm error && error.description() instanceof MetaCall hole) {
     //   hole.ref().conditions.append(Tuple.of(lhsSubst, rhsTerm));
@@ -50,7 +55,7 @@ public record YouTrack(
     // if (rhsTerm instanceof ErrorTerm error && error.description() instanceof MetaCall hole) {
     //   hole.ref().conditions.append(Tuple.of(rhsSubst, lhsTerm));
     // }
-    result = tycker.whnf(result.instantiateTele(args.view()));
+    result = tycker.whnf(result.instantiateTele(unifyResult.args().view()));
     var old = tycker.setLocalCtx(ctx);
     tycker.unifyTermReported(lhsTerm, rhsTerm, result, pos, comparison ->
       new ClausesProblem.Confluence(pos, rhsInfo.ix + 1, lhsInfo.ix + 1,
@@ -58,15 +63,13 @@ public record YouTrack(
     tycker.setLocalCtx(old);
   }
 
-  // TODO: this implementation is incorrect. To do it correctly, we will need to do another unification,
-  //  and I am unsure if it is worth the effort.
   private void domination(
-    LocalCtx ctx, ImmutableSeq<Term> subst,
+    LocalCtx ctx, Seq<Term> subst,
     int lhsIx, int rhsIx, Term.Matching matching,
     MutableSet<ClausesProblem.Domination> doms
   ) {
-    // if (subst.allMatch(dom -> dom instanceof FreeTerm(var ref) && ctx.contains(ref)))
-    //   doms.add(new ClausesProblem.Domination(lhsIx + 1, rhsIx + 1, matching.sourcePos()));
+    if (subst.allMatch(dom -> dom instanceof FreeTerm(var ref) && ctx.contains(ref)))
+      doms.add(new ClausesProblem.Domination(lhsIx + 1, rhsIx + 1, matching.sourcePos()));
   }
 
   public void check(

--- a/base/src/main/java/org/aya/tyck/tycker/AppTycker.java
+++ b/base/src/main/java/org/aya/tyck/tycker/AppTycker.java
@@ -101,13 +101,12 @@ public interface AppTycker {
           var ownerArgs = realArgs.take(ownerTele.size());
           var conArgs = realArgs.drop(ownerTele.size());
 
-          var type = fullSignature.result(realArgs);
-          var dataCall = new DataCall(dataVar, 0, ownerArgs);
+          var type = (DataCall) fullSignature.result(realArgs);
           var shape = state.shapeFactory().find(new DataDef.Delegate(dataVar))
-            .mapNotNull(recog -> AyaShape.ofCon(new ConDef.Delegate(conVar), recog, dataCall))
+            .mapNotNull(recog -> AyaShape.ofCon(new ConDef.Delegate(conVar), recog, type))
             .getOrNull();
-          if (shape != null) return new Jdg.Default(new RuleReducer.Con(shape, 0, ownerArgs, conArgs), type);
-          var wellTyped = new ConCall(conVar, 0, ownerArgs, conArgs);
+          if (shape != null) return new Jdg.Default(new RuleReducer.Con(shape, 0, type.args(), conArgs), type);
+          var wellTyped = new ConCall(conVar, 0, type.args(), conArgs);
           return new Jdg.Default(wellTyped, type);
         });
       }

--- a/base/src/main/java/org/aya/unify/Synthesizer.java
+++ b/base/src/main/java/org/aya/unify/Synthesizer.java
@@ -10,6 +10,7 @@ import org.aya.syntax.core.def.PrimDef;
 import org.aya.syntax.core.def.TyckDef;
 import org.aya.syntax.core.term.*;
 import org.aya.syntax.core.term.call.Callable;
+import org.aya.syntax.core.term.call.ConCall;
 import org.aya.syntax.core.term.call.MetaCall;
 import org.aya.syntax.core.term.repr.IntegerTerm;
 import org.aya.syntax.core.term.repr.ListTerm;
@@ -113,6 +114,7 @@ public record Synthesizer(
       }
       case IntegerTerm lit -> lit.type();
       case ListTerm list -> list.type();
+      case ConCall conCall -> conCall.head().underlyingDataCall();
       case Callable.Tele teleCall -> TyckDef.defSignature(teleCall.ref())
         .result(teleCall.args())
         .elevate(teleCall.ulift());

--- a/base/src/main/java/org/aya/unify/TermComparator.java
+++ b/base/src/main/java/org/aya/unify/TermComparator.java
@@ -10,6 +10,7 @@ import org.aya.generic.term.SortKind;
 import org.aya.prettier.AyaPrettierOptions;
 import org.aya.syntax.compile.JitTele;
 import org.aya.syntax.core.def.AnyDef;
+import org.aya.syntax.core.def.ConDefLike;
 import org.aya.syntax.core.def.TyckDef;
 import org.aya.syntax.core.term.*;
 import org.aya.syntax.core.term.call.*;
@@ -92,9 +93,20 @@ public abstract sealed class TermComparator extends AbstractTycker permits Unifi
       case Pair(PrimCall lFn, PrimCall rFn) -> compareCallApprox(lFn, rFn, lFn.ref());
       case Pair(IntegerTerm lInt, IntegerTerm rInt) ->
         lInt.repr() == ((Shaped.@NotNull Nat<Term>) rInt).repr() ? lInt.type() : null;
-      case Pair(ConCallLike lCon, ConCallLike rCon) -> compareCallApprox(lCon, rCon, lCon.ref());
+      case Pair(ConCallLike lCon, ConCallLike rCon) -> lossyUnifyCon(lCon, rCon, lCon.ref());
       default -> null;
     };
+  }
+  private @Nullable Term lossyUnifyCon(ConCallLike lCon, ConCallLike rCon, @NotNull ConDefLike ref) {
+    if (!lCon.ref().equals(rCon.ref()) || lCon.ulift() != rCon.ulift()) return null;
+    // since dataArgs live in a separate telescope
+    if (null == compareMany(lCon.head().ownerArgs(),
+      rCon.head().ownerArgs(), lCon.ulift(), TyckDef.defSignature(ref.dataRef())))
+      return null;
+    if (null == compareMany(lCon.conArgs(), rCon.conArgs(), lCon.ulift(),
+      new JitTele.LocallyNameless(ref.selfTele(lCon.args()), SortTerm.Type0)))
+      return null;
+    return new Synthesizer(this).synth(lCon);
   }
 
   /**
@@ -103,7 +115,7 @@ public abstract sealed class TermComparator extends AbstractTycker permits Unifi
   private @Nullable Term compareCallApprox(
     @NotNull Callable.Tele lhs, @NotNull Callable.Tele rhs, @NotNull AnyDef typeProvider
   ) {
-    if (!lhs.ref().equals(rhs.ref())) return null;
+    if (!lhs.ref().equals(rhs.ref()) || lhs.ulift() != rhs.ulift()) return null;
     return compareMany(lhs.args(), rhs.args(), lhs.ulift(), TyckDef.defSignature(typeProvider));
   }
 
@@ -289,7 +301,7 @@ public abstract sealed class TermComparator extends AbstractTycker permits Unifi
       // fallback case
       case ConCallLike lCon -> switch (rhs) {
         case ListTerm rList -> compareUntyped(lhs, rList.constructorForm());
-        case ConCallLike rCon -> compareCallApprox(lCon, rCon, lCon.ref());
+        case ConCallLike rCon -> lossyUnifyCon(lCon, rCon, lCon.ref());
         default -> null;
       };
       case MetaLitTerm mlt -> switch (rhs) {

--- a/base/src/main/java/org/aya/unify/TermComparator.java
+++ b/base/src/main/java/org/aya/unify/TermComparator.java
@@ -9,8 +9,6 @@ import org.aya.generic.stmt.Shaped;
 import org.aya.generic.term.SortKind;
 import org.aya.prettier.AyaPrettierOptions;
 import org.aya.syntax.compile.JitTele;
-import org.aya.syntax.core.def.AnyDef;
-import org.aya.syntax.core.def.ConDefLike;
 import org.aya.syntax.core.def.TyckDef;
 import org.aya.syntax.core.term.*;
 import org.aya.syntax.core.term.call.*;
@@ -89,34 +87,35 @@ public abstract sealed class TermComparator extends AbstractTycker permits Unifi
    */
   private @Nullable Term compareApprox(@NotNull Term lhs, @NotNull Term rhs) {
     return switch (new Pair<>(lhs, rhs)) {
-      case Pair(FnCall lFn, FnCall rFn) -> compareCallApprox(lFn, rFn, lFn.ref());
-      case Pair(PrimCall lFn, PrimCall rFn) -> compareCallApprox(lFn, rFn, lFn.ref());
+      case Pair(FnCall lFn, FnCall rFn) -> compareCallApprox(lFn, rFn);
+      case Pair(PrimCall lFn, PrimCall rFn) -> compareCallApprox(lFn, rFn);
       case Pair(IntegerTerm lInt, IntegerTerm rInt) ->
         lInt.repr() == ((Shaped.@NotNull Nat<Term>) rInt).repr() ? lInt.type() : null;
-      case Pair(ConCallLike lCon, ConCallLike rCon) -> lossyUnifyCon(lCon, rCon, lCon.ref());
+      case Pair(ConCallLike lCon, ConCallLike rCon) -> lossyUnifyCon(lCon, rCon);
       default -> null;
     };
   }
-  private @Nullable Term lossyUnifyCon(ConCallLike lCon, ConCallLike rCon, @NotNull ConDefLike ref) {
+  private @Nullable Term lossyUnifyCon(ConCallLike lCon, ConCallLike rCon) {
     if (!lCon.ref().equals(rCon.ref()) || lCon.ulift() != rCon.ulift()) return null;
     // since dataArgs live in a separate telescope
-    if (null == compareMany(lCon.head().ownerArgs(),
-      rCon.head().ownerArgs(), lCon.ulift(), TyckDef.defSignature(ref.dataRef())))
+    var lHead = lCon.head();
+    if (null == compareMany(lHead.ownerArgs(),
+      rCon.head().ownerArgs(), lHead.ulift(), TyckDef.defSignature(lHead.ref().dataRef())))
       return null;
-    if (null == compareMany(lCon.conArgs(), rCon.conArgs(), lCon.ulift(),
-      new JitTele.LocallyNameless(ref.selfTele(lCon.args()), SortTerm.Type0)))
+    if (null == compareMany(lCon.conArgs(), rCon.conArgs(), lHead.ulift(),
+      new JitTele.LocallyNameless(lHead.ref().selfTele(lCon.args()), SortTerm.Type0)))
       return null;
-    return new Synthesizer(this).synth(lCon);
+    return lHead.underlyingDataCall();
   }
 
   /**
    * Compare the arguments of two callable ONLY, this method will NOT try to normalize and then compare (while the old project does).
    */
   private @Nullable Term compareCallApprox(
-    @NotNull Callable.Tele lhs, @NotNull Callable.Tele rhs, @NotNull AnyDef typeProvider
+    @NotNull Callable.Tele lhs, @NotNull Callable.Tele rhs
   ) {
     if (!lhs.ref().equals(rhs.ref()) || lhs.ulift() != rhs.ulift()) return null;
-    return compareMany(lhs.args(), rhs.args(), lhs.ulift(), TyckDef.defSignature(typeProvider));
+    return compareMany(lhs.args(), rhs.args(), lhs.ulift(), TyckDef.defSignature(lhs.ref()));
   }
 
   private <R> R swapped(@NotNull Supplier<R> callback) {
@@ -301,7 +300,7 @@ public abstract sealed class TermComparator extends AbstractTycker permits Unifi
       // fallback case
       case ConCallLike lCon -> switch (rhs) {
         case ListTerm rList -> compareUntyped(lhs, rList.constructorForm());
-        case ConCallLike rCon -> lossyUnifyCon(lCon, rCon, lCon.ref());
+        case ConCallLike rCon -> lossyUnifyCon(lCon, rCon);
         default -> null;
       };
       case MetaLitTerm mlt -> switch (rhs) {

--- a/base/src/main/java/org/aya/unify/TermComparator.java
+++ b/base/src/main/java/org/aya/unify/TermComparator.java
@@ -343,7 +343,7 @@ public abstract sealed class TermComparator extends AbstractTycker permits Unifi
     assert rist.sizeEquals(types.telescopeSize);
     var argsCum = new Term[types.telescopeSize];
 
-    for (var i = 0; i < list.size(); ++i) {
+    for (var i = 0; i < types.telescopeSize; ++i) {
       var l = list.get(i);
       var r = rist.get(i);
       var ty = whnf(types.telescope(i, argsCum)).elevate(ulift);

--- a/cli-console/src/main/java/org/aya/cli/console/AnsiReporter.java
+++ b/cli-console/src/main/java/org/aya/cli/console/AnsiReporter.java
@@ -18,8 +18,6 @@ import java.util.function.Supplier;
  * @author ice1000
  */
 public record AnsiReporter(
-  // TODO: jansi is deprecated, we use jline3-jni now,
-  //  how can we find a replacement?
   boolean supportAnsi,
   @NotNull BooleanSupplier unicode,
   @NotNull Supplier<PrettierOptions> options,

--- a/cli-console/src/main/java/org/aya/cli/repl/jline/JlineRepl.java
+++ b/cli-console/src/main/java/org/aya/cli/repl/jline/JlineRepl.java
@@ -1,4 +1,4 @@
-// Copyright (c) 2020-2023 Tesla (Yinsen) Zhang.
+// Copyright (c) 2020-2024 Tesla (Yinsen) Zhang.
 // Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
 package org.aya.cli.repl.jline;
 
@@ -44,7 +44,7 @@ public final class JlineRepl extends AyaRepl {
   public JlineRepl(@NotNull ImmutableSeq<Path> modulePaths, @NotNull ReplConfig config) throws IOException {
     super(modulePaths, config);
     terminal = TerminalBuilder.builder()
-      .jansi(true)
+      .jni(true)
       .jna(false)
       .build();
     lineReader = LineReaderBuilder.builder()

--- a/cli-impl/src/test/resources/negative/PatCohError.txt
+++ b/cli-impl/src/test/resources/negative/PatCohError.txt
@@ -154,7 +154,77 @@ Error: The 4th and the 3rd clauses are not confluent because we failed to unify
        and
          suc (suc a)
 
-6 error(s), 0 warning(s).
+In file $FILE:3:2 ->
+
+  1 │   open import arith::Nat
+  2 │   overlap def test Nat : Nat
+  3 │   | 0 => 0
+    │     ╰────╯
+
+Warning: The 2nd clause dominates the 1st clause. The 1st clause will be 
+         unreachable
+
+In file $FILE:5:2 ->
+
+  3 │   | 0 => 0
+  4 │   | a => a
+  5 │   | suc a => suc a
+    │     ╰────────────╯
+
+Warning: The 2nd clause dominates the 3rd clause. The 3rd clause will be 
+         unreachable
+
+In file $FILE:9:2 ->
+
+  7 │   | 2147483647 => 3
+  8 │   | 2147483647 => 4
+  9 │   | 114514 => 1919
+    │     ╰────────────╯
+
+Warning: The 2nd clause dominates the 7th clause. The 7th clause will be 
+         unreachable
+
+In file $FILE:6:2 ->
+
+  4 │   | a => a
+  5 │   | suc a => suc a
+  6 │   | suc (suc a) => a
+    │     ╰──────────────╯
+
+Warning: The 3rd clause dominates the 4th clause. The 4th clause will be 
+         unreachable
+
+In file $FILE:8:2 ->
+
+  6 │   | suc (suc a) => a
+  7 │   | 2147483647 => 3
+  8 │   | 2147483647 => 4
+    │     ╰─────────────╯
+
+Warning: The 5th clause dominates the 6th clause. The 6th clause will be 
+         unreachable
+
+In file $FILE:7:2 ->
+
+  5 │   | suc a => suc a
+  6 │   | suc (suc a) => a
+  7 │   | 2147483647 => 3
+    │     ╰─────────────╯
+
+Warning: The 6th clause dominates the 5th clause. The 5th clause will be 
+         unreachable
+
+In file $FILE:8:2 ->
+
+  6 │   | suc (suc a) => a
+  7 │   | 2147483647 => 3
+  8 │   | 2147483647 => 4
+    │     ╰─────────────╯
+
+Warning: The 2nd clause dominates the 6th clause. The 6th clause will be 
+         unreachable
+
+6 error(s), 7 warning(s).
 What are you doing?
 
 ConflLiteral2:
@@ -174,6 +244,16 @@ Error: The 1st and the 2nd clauses are not confluent because we failed to unify
        and
          1
 
+In file $FILE:4:2 ->
+
+  2 │   overlap def largeInt1 Nat Nat : Nat
+  3 │   | a, b => a
+  4 │   | 114514, 1919810 => 1
+    │     ╰──────────────────╯
+
+Warning: The 1st clause dominates the 2nd clause. The 2nd clause will be 
+         unreachable
+
 In file $FILE:6:12 ->
 
   4 │       | 114514, 1919810 => 1
@@ -189,6 +269,16 @@ Error: The 1st and the 2nd clauses are not confluent because we failed to unify
          1919810
        and
          1
+
+In file $FILE:8:2 ->
+
+  6 │   overlap def largeInt2 Nat Nat : Nat
+  7 │   | a, b => b
+  8 │   | a, 1919810 => 1
+    │     ╰─────────────╯
+
+Warning: The 1st clause dominates the 2nd clause. The 2nd clause will be 
+         unreachable
 
 In file $FILE:10:12 ->
 
@@ -239,6 +329,26 @@ Error: The 2nd and the 1st clauses are not confluent because we failed to unify
        and
          suc b
 
+In file $FILE:13:2 ->
+
+  11 │   | a, b => b
+  12 │   | a, suc b => b
+  13 │   | a, 1919810 => 1
+     │     ╰─────────────╯
+
+Warning: The 1st clause dominates the 3rd clause. The 3rd clause will be 
+         unreachable
+
+In file $FILE:12:2 ->
+
+  10 │   overlap def largeInt3 Nat Nat : Nat
+  11 │   | a, b => b
+  12 │   | a, suc b => b
+     │     ╰───────────╯
+
+Warning: The 1st clause dominates the 2nd clause. The 2nd clause will be 
+         unreachable
+
 In file $FILE:15:12 ->
 
   13 │       | a, 1919810 => 1
@@ -255,6 +365,16 @@ Error: The 1st and the 2nd clauses are not confluent because we failed to unify
        and
          1
 
+In file $FILE:17:2 ->
+
+  15 │   overlap def largeInt1-inv Nat Nat : Nat
+  16 │   | a, b => a
+  17 │   | 114514, 1919810 => 1
+     │     ╰──────────────────╯
+
+Warning: The 1st clause dominates the 2nd clause. The 2nd clause will be 
+         unreachable
+
 In file $FILE:19:12 ->
 
   17 │       | 114514, 1919810 => 1
@@ -270,6 +390,16 @@ Error: The 1st and the 2nd clauses are not confluent because we failed to unify
          1919810
        and
          1
+
+In file $FILE:21:2 ->
+
+  19 │   overlap def largeInt2-inv Nat Nat : Nat
+  20 │   | b, a => b
+  21 │   | 1919810, a => 1
+     │     ╰─────────────╯
+
+Warning: The 1st clause dominates the 2nd clause. The 2nd clause will be 
+         unreachable
 
 In file $FILE:23:12 ->
 
@@ -319,6 +449,26 @@ Error: The 2nd and the 1st clauses are not confluent because we failed to unify
          b
        and
          suc b
+
+In file $FILE:26:2 ->
+
+  24 │   | b, a => b
+  25 │   | suc b, a => b
+  26 │   | 1919810, a => 1
+     │     ╰─────────────╯
+
+Warning: The 1st clause dominates the 3rd clause. The 3rd clause will be 
+         unreachable
+
+In file $FILE:25:2 ->
+
+  23 │   overlap def largeInt3-inv Nat Nat : Nat
+  24 │   | b, a => b
+  25 │   | suc b, a => b
+     │     ╰───────────╯
+
+Warning: The 1st clause dominates the 2nd clause. The 2nd clause will be 
+         unreachable
 
 In file $FILE:28:12 ->
 
@@ -338,7 +488,47 @@ Error: The 2nd and the 3rd clauses are not confluent because we failed to unify
        and
          515
 
-11 error(s), 0 warning(s).
+In file $FILE:29:2 ->
+
+  27 │   
+  28 │   overlap def multi-nodes Nat Nat : Nat
+  29 │   | 114, 0 => 0
+     │     ╰─────────╯
+
+Warning: The 5th clause dominates the 1st clause. The 1st clause will be 
+         unreachable
+
+In file $FILE:31:2 ->
+
+  29 │   | 114, 0 => 0
+  30 │   | 114, suc b => suc b
+  31 │   | 114, 514 => 515
+     │     ╰─────────────╯
+
+Warning: The 2nd clause dominates the 3rd clause. The 3rd clause will be 
+         unreachable
+
+In file $FILE:30:2 ->
+
+  28 │   overlap def multi-nodes Nat Nat : Nat
+  29 │   | 114, 0 => 0
+  30 │   | 114, suc b => suc b
+     │     ╰─────────────────╯
+
+Warning: The 5th clause dominates the 2nd clause. The 2nd clause will be 
+         unreachable
+
+In file $FILE:32:2 ->
+
+  30 │   | 114, suc b => suc b
+  31 │   | 114, 514 => 515
+  32 │   | 115, 514 => 514
+     │     ╰─────────────╯
+
+Warning: The 5th clause dominates the 4th clause. The 4th clause will be 
+         unreachable
+
+11 error(s), 12 warning(s).
 What are you doing?
 
 FirstMatchDomination:
@@ -388,7 +578,47 @@ Error: The 5th and the 3rd clauses are not confluent because we failed to unify
        and
          1
 
-2 error(s), 0 warning(s).
+In file $FILE:4:2 ->
+
+  2 │   overlap def not-conf Nat : Nat
+  3 │   | zero => 1
+  4 │   | 0 => 0
+    │     ╰────╯
+
+Warning: The 1st clause dominates the 2nd clause. The 2nd clause will be 
+         unreachable
+
+In file $FILE:3:2 ->
+
+  1 │   open import arith::Nat
+  2 │   overlap def not-conf Nat : Nat
+  3 │   | zero => 1
+    │     ╰───────╯
+
+Warning: The 2nd clause dominates the 1st clause. The 1st clause will be 
+         unreachable
+
+In file $FILE:5:2 ->
+
+  3 │   | zero => 1
+  4 │   | 0 => 0
+  5 │   | 1 => 1
+    │     ╰────╯
+
+Warning: The 5th clause dominates the 3rd clause. The 3rd clause will be 
+         unreachable
+
+In file $FILE:6:2 ->
+
+  4 │   | 0 => 0
+  5 │   | 1 => 1
+  6 │   | suc 1 => 1
+    │     ╰────────╯
+
+Warning: The 5th clause dominates the 4th clause. The 4th clause will be 
+         unreachable
+
+2 error(s), 4 warning(s).
 What are you doing?
 
 NestedMissing:

--- a/cli-impl/src/test/resources/shared/src/data/Interval.aya
+++ b/cli-impl/src/test/resources/shared/src/data/Interval.aya
@@ -14,3 +14,6 @@ private def lemma
 | left ⇒ f a
 | right ⇒ g a
 | line j ⇒ p a j
+
+example def funExt' (f g : A -> B) (p : ∀ a -> f a = g a) : f = g =>
+  pmap (lemma f g p) (fn i => line i)

--- a/cli-impl/src/test/resources/shared/src/data/Vec.aya
+++ b/cli-impl/src/test/resources/shared/src/data/Vec.aya
@@ -1,4 +1,5 @@
 open import arith::Nat
+open import Paths
 
 open data Vec Nat Type
 | 0, A => []
@@ -16,3 +17,9 @@ def head (Vec (suc n) A) : A
 
 def tail (Vec (suc n) A) : Vec n A
 | _ :> xs => xs
+
+overlap def infixr ++ (Vec n A) (Vec m A) : Vec (n + m) A
+| [], ys => ys
+| ys, [] => ys
+| x :> xs, ys => x :> xs ++ ys
+tighter :> =

--- a/jit-compiler/src/main/java/org/aya/compiler/PatternSerializer.java
+++ b/jit-compiler/src/main/java/org/aya/compiler/PatternSerializer.java
@@ -198,9 +198,9 @@ public final class PatternSerializer extends AbstractSerializer<ImmutableSeq<Pat
     appendLine(STR."\{VARIABLE_RESULT}.set(\{bindCount++}, \{term});");
   }
 
-  private int bindAmount(@NotNull Pat pat) {
+  private int bindAmount(@NotNull ImmutableSeq<Pat> pats) {
     var acc = MutableIntValue.create();
-    pat.consumeBindings((_, _) -> acc.increment());
+    pats.forEach(pat -> pat.consumeBindings((_, _) -> acc.increment()));
     return acc.get();
   }
 
@@ -208,8 +208,7 @@ public final class PatternSerializer extends AbstractSerializer<ImmutableSeq<Pat
 
   @Override
   public AyaSerializer<ImmutableSeq<Matching>> serialize(@NotNull ImmutableSeq<Matching> unit) {
-    var bindSize = unit.mapToInt(ImmutableIntSeq.factory(),
-      x -> x.patterns.view().foldLeft(0, (acc, p) -> acc + bindAmount(p)));
+    var bindSize = unit.mapToInt(ImmutableIntSeq.factory(), x -> bindAmount(x.patterns));
     int maxBindSize = bindSize.max();
 
     buildLocalVar(STR."\{CLASS_MUTSEQ}<\{CLASS_TERM}>", VARIABLE_RESULT, STR."\{CLASS_MUTSEQ}.fill(\{maxBindSize}, (\{CLASS_TERM}) null)");

--- a/syntax/src/main/java/org/aya/syntax/compile/JitCon.java
+++ b/syntax/src/main/java/org/aya/syntax/compile/JitCon.java
@@ -5,6 +5,7 @@ package org.aya.syntax.compile;
 import kala.collection.Seq;
 import kala.collection.immutable.ImmutableSeq;
 import kala.collection.mutable.MutableArrayList;
+import kala.collection.mutable.MutableList;
 import kala.control.Result;
 import org.aya.syntax.core.def.ConDefLike;
 import org.aya.syntax.core.def.DataDefLike;
@@ -44,10 +45,9 @@ public abstract non-sealed class JitCon extends JitDef implements ConDefLike {
   @Override public @NotNull DataDefLike dataRef() { return dataType; }
 
   @Override public @NotNull ImmutableSeq<Param> selfTele(@NotNull ImmutableSeq<Term> ownerArgs) {
-    var ownerArgsSize = ownerArgs.size();
+    var args = MutableList.from(isAvailable(ownerArgs).get());
+    var ownerArgsSize = args.size();
     var selfArgsSize = telescopeSize - ownerArgsSize;
-    var args = MutableArrayList.<Term>create(telescopeSize);
-    args.appendAll(ownerArgs);
     var tele = MutableArrayList.<Param>create(selfArgsSize);
 
     for (var i = 0; i < selfArgsSize; ++i) {

--- a/syntax/src/main/java/org/aya/syntax/core/pat/Pat.java
+++ b/syntax/src/main/java/org/aya/syntax/core/pat/Pat.java
@@ -13,12 +13,14 @@ import org.aya.prettier.CorePrettier;
 import org.aya.prettier.Tokens;
 import org.aya.pretty.doc.Doc;
 import org.aya.syntax.core.def.ConDefLike;
+import org.aya.syntax.core.term.Param;
 import org.aya.syntax.core.term.Term;
 import org.aya.syntax.core.term.call.DataCall;
 import org.aya.syntax.core.term.repr.IntegerTerm;
 import org.aya.syntax.ref.GenerateKind;
 import org.aya.syntax.ref.LocalCtx;
 import org.aya.syntax.ref.LocalVar;
+import org.aya.util.Pair;
 import org.aya.util.error.Panic;
 import org.aya.util.error.SourcePos;
 import org.aya.util.error.WithPos;
@@ -44,13 +46,18 @@ public sealed interface Pat extends AyaDocile {
    * The order of bindings should be postorder, that is, {@code (Con0 a (Con1 b)) as c} should be {@code [a , b , c]}
    */
   void consumeBindings(@NotNull BiConsumer<LocalVar, Term> consumer);
+  @NotNull Pat bind(MutableList<LocalVar> vars);
 
-  record CollectBind(LocalVar var, Term type) { }
-
-  static @NotNull MutableList<CollectBind> collectBindings(@NotNull SeqView<Pat> pats) {
-    var buffer = MutableList.<CollectBind>create();
+  static @NotNull Pair<MutableList<LocalVar>, ImmutableSeq<Pat>>
+  collectVariables(@NotNull SeqView<Pat> pats) {
+    var buffer = MutableList.<LocalVar>create();
+    var newPats = pats.map(p -> p.bind(buffer)).toImmutableSeq();
+    return new Pair<>(buffer, newPats);
+  }
+  static @NotNull MutableList<Param> collectBindings(@NotNull SeqView<Pat> pats) {
+    var buffer = MutableList.<Param>create();
     pats.forEach(p -> p.consumeBindings((var, type) ->
-      buffer.append(new CollectBind(var, type))));
+      buffer.append(new Param(var.name(), type, false))));
     return buffer;
   }
 
@@ -67,6 +74,7 @@ public sealed interface Pat extends AyaDocile {
     }
 
     @Override public void consumeBindings(@NotNull BiConsumer<LocalVar, Term> consumer) { }
+    @Override public @NotNull Pat bind(MutableList<LocalVar> vars) { return this; }
     @Override public @NotNull Pat inline(@NotNull BiConsumer<LocalVar, Term> bind) { return this; }
   }
 
@@ -86,6 +94,11 @@ public sealed interface Pat extends AyaDocile {
     @Override public void consumeBindings(@NotNull BiConsumer<LocalVar, Term> consumer) {
       consumer.accept(bind, type);
     }
+    @Override public @NotNull Pat bind(MutableList<LocalVar> vars) {
+      var newType = type.bindTele(vars.view());
+      vars.append(bind);
+      return update(newType);
+    }
 
     @Override public @NotNull Pat inline(@NotNull BiConsumer<LocalVar, Term> bind) { return this; }
   }
@@ -95,13 +108,15 @@ public sealed interface Pat extends AyaDocile {
       return this.elements.sameElements(elements, true) ? this : new Tuple(elements);
     }
 
-    @Override
-    public @NotNull Pat descent(@NotNull UnaryOperator<Pat> patOp, @NotNull UnaryOperator<Term> termOp) {
+    @Override public @NotNull Pat descent(@NotNull UnaryOperator<Pat> patOp, @NotNull UnaryOperator<Term> termOp) {
       return update(elements.map(patOp));
     }
 
     @Override public void consumeBindings(@NotNull BiConsumer<LocalVar, Term> consumer) {
       elements.forEach(e -> e.consumeBindings(consumer));
+    }
+    @Override public @NotNull Pat bind(MutableList<LocalVar> vars) {
+      return update(elements.map(e -> e.bind(vars)));
     }
 
     @Override public @NotNull Pat inline(@NotNull BiConsumer<LocalVar, Term> bind) {
@@ -124,6 +139,10 @@ public sealed interface Pat extends AyaDocile {
     }
     @Override public void consumeBindings(@NotNull BiConsumer<LocalVar, Term> consumer) {
       args.forEach(e -> e.consumeBindings(consumer));
+    }
+    @Override public @NotNull Pat bind(MutableList<LocalVar> vars) {
+      var newData = (DataCall) data.bindTele(vars.view());
+      return update(args.map(e -> e.bind(vars)), newData);
     }
 
     @Override public @NotNull Pat inline(@NotNull BiConsumer<LocalVar, Term> bind) {
@@ -157,8 +176,12 @@ public sealed interface Pat extends AyaDocile {
     }
 
     @Override public void consumeBindings(@NotNull BiConsumer<LocalVar, Term> consumer) {
-      // We should call storeBindings after inline
+      // Called after inline
       Panic.unreachable();
+    }
+    @Override public @NotNull Pat bind(MutableList<LocalVar> vars) {
+      // Called after inline
+      return Panic.unreachable();
     }
 
     @Override public @NotNull Pat inline(@NotNull BiConsumer<LocalVar, Term> bind) {
@@ -195,6 +218,9 @@ public sealed interface Pat extends AyaDocile {
       return this;
     }
     @Override public void consumeBindings(@NotNull BiConsumer<LocalVar, Term> consumer) { }
+    @Override public @NotNull Pat bind(MutableList<LocalVar> vars) {
+      return update((DataCall) type.bindTele(vars.view()));
+    }
     @Override public @NotNull Con makeZero() {
       return new Pat.Con(zero, ImmutableSeq.empty(), type);
     }

--- a/syntax/src/main/java/org/aya/syntax/core/pat/Pat.java
+++ b/syntax/src/main/java/org/aya/syntax/core/pat/Pat.java
@@ -15,6 +15,7 @@ import org.aya.pretty.doc.Doc;
 import org.aya.syntax.core.def.ConDefLike;
 import org.aya.syntax.core.term.Param;
 import org.aya.syntax.core.term.Term;
+import org.aya.syntax.core.term.call.ConCallLike;
 import org.aya.syntax.core.term.call.DataCall;
 import org.aya.syntax.core.term.repr.IntegerTerm;
 import org.aya.syntax.ref.GenerateKind;
@@ -147,6 +148,9 @@ public sealed interface Pat extends AyaDocile {
 
     @Override public @NotNull Pat inline(@NotNull BiConsumer<LocalVar, Term> bind) {
       return update(args.map(x -> x.inline(bind)), data);
+    }
+    public @NotNull ConCallLike.Head conHead() {
+      return new ConCallLike.Head(ref, 0, data.args());
     }
   }
 

--- a/syntax/src/main/java/org/aya/syntax/core/pat/PatToTerm.java
+++ b/syntax/src/main/java/org/aya/syntax/core/pat/PatToTerm.java
@@ -6,14 +6,11 @@ import kala.collection.SeqView;
 import kala.collection.immutable.ImmutableSeq;
 import org.aya.syntax.core.term.*;
 import org.aya.syntax.core.term.call.ConCall;
-import org.aya.syntax.core.term.call.ConCallLike;
 import org.aya.syntax.core.term.xtt.DimTerm;
 import org.aya.syntax.ref.LocalCtx;
-import org.aya.util.Pair;
 import org.aya.util.error.Panic;
 import org.jetbrains.annotations.NotNull;
 
-import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
 
@@ -30,36 +27,10 @@ public interface PatToTerm {
           freshCallback.accept(bind);
           yield new FreeTerm(bind.bind());
         }
-        case Pat.Con con -> new ConCall(conHead(con), con.args().map(this));
+        case Pat.Con con -> new ConCall(con.conHead(), con.args().map(this));
         case Pat.Tuple tuple -> new TupTerm(tuple.elements().map(this));
         case Pat.Meta meta -> new MetaPatTerm(meta);
         case Pat.ShapedInt si -> si.toTerm();
-      };
-    }
-  }
-  private static ConCallLike.@NotNull Head conHead(Pat.Con con) {
-    return new ConCallLike.Head(con.ref(), 0, con.data().args());
-  }
-
-  record Binary(@NotNull LocalCtx ctx, @NotNull Unary unary) implements BiFunction<Pat, Pat, Term> {
-    public Binary(@NotNull LocalCtx ctx) {
-      this(ctx, new Unary(bind -> ctx.put(bind.bind(), bind.type())));
-    }
-    public @NotNull ImmutableSeq<Term> list(@NotNull ImmutableSeq<Pat> lhs, @NotNull ImmutableSeq<Pat> rhs) {
-      return lhs.zip(rhs, this);
-    }
-
-    public @NotNull Term apply(@NotNull Pat l, @NotNull Pat r) {
-      return switch (new Pair<>(l, r)) {
-        case Pair(Pat.Bind _, var rhs) -> unary.apply(rhs);
-        case Pair(var lhs, Pat.Bind _) -> unary.apply(lhs);
-        // It must be the case that lhs.ref == rhs.ref
-        case Pair(Pat.Con lhs, Pat.Con rhs) -> new ConCall(conHead(lhs), list(lhs.args(), rhs.args()));
-        case Pair(Pat.ShapedInt lhs, Pat.Con rhs) -> apply(lhs.constructorForm(), rhs);
-        case Pair(Pat.Con lhs, Pat.ShapedInt rhs) -> apply(lhs, rhs.constructorForm());
-        case Pair(Pat.ShapedInt lhs, Pat.ShapedInt _) -> lhs.toTerm();
-        case Pair(Pat.Tuple lhs, Pat.Tuple rhs) -> new TupTerm(list(lhs.elements(), rhs.elements()));
-        default -> Panic.unreachable();
       };
     }
   }
@@ -95,9 +66,9 @@ public interface PatToTerm {
         }
         case Pat.Con con when con.ref().hasEq() ->
           list(con.args().view().dropLast(1), BOUNDARIES)
-            .map(args -> new ConCall(conHead(con), args));
+            .map(args -> new ConCall(con.conHead(), args));
         case Pat.Con con -> list(con.args().view())
-          .map(args -> new ConCall(conHead(con), args));
+          .map(args -> new ConCall(con.conHead(), args));
         case Pat.Tuple tuple -> list(tuple.elements().view())
           .map(args -> new TupTerm(args));
       };

--- a/syntax/src/main/java/org/aya/syntax/core/term/call/ConCallLike.java
+++ b/syntax/src/main/java/org/aya/syntax/core/term/call/ConCallLike.java
@@ -32,6 +32,10 @@ public sealed interface ConCallLike extends Callable.Tele permits ConCall, RuleR
       if (args.sameElements(ownerArgs, true)) return this;
       return new Head(ref, ulift, args);
     }
+
+    public @NotNull DataCall underlyingDataCall() {
+      return new DataCall(ref.dataRef(), ulift, ownerArgs);
+    }
   }
 
   @NotNull ConCallLike.Head head();

